mscore/masterpalette.cpp

```diff
@@ -263,5 +265,114 @@ void MasterPalette::changeEvent(QEvent *event)
             retranslate();
       }
 
+//---------------------------------------------------------
+//   keyPressEvent
+//---------------------------------------------------------
+
+void MasterPalette::keyPressEvent(QKeyEvent *event)
+      {
+      QTreeWidgetItem* currentItem = treeWidget->currentItem();
+      int idx = treeWidget->indexOfTopLevelItem(currentItem);
+      if (event->key() == Qt::Key_Down) {
+            if (idx >= 0 && idx < treeWidget->topLevelItemCount() - 1) {
+                  selectItem(treeWidget->topLevelItem(idx+1)->text(0));
+                  }
+            else if (currentItem == symbolItem) {
+                  if (symbolItem->isExpanded())
+                        treeWidget->setCurrentItem(symbolItem->child(0));
+                  else
+                        selectItem(treeWidget->topLevelItem(0)->text(0));
+                  }
+            else if (symbolItem->indexOfChild(currentItem) != -1) {
+                  int i = symbolItem->indexOfChild(currentItem);
+                  if (i < symbolItem->childCount()-1) {
+                        treeWidget->setCurrentItem(symbolItem->child(i+1));
+                        }
+                  else {
+                        int symIdx = treeWidget->indexOfTopLevelItem(symbolItem);
+                        if (symIdx < treeWidget->topLevelItemCount() - 1)
+                              selectItem(treeWidget->topLevelItem(symIdx+1)->text(0));
+                        else
+                              selectItem(treeWidget->topLevelItem(0)->text(0));
+                        }
+                  }
+            else {
+                  selectItem(treeWidget->topLevelItem(0)->text(0));
+                  }
+            }
+      else if (event->key() == Qt::Key_Up) {
+            if (idx > 0) {
+                  selectItem(treeWidget->topLevelItem(idx-1)->text(0));
+                  }
+            else if (symbolItem->indexOfChild(currentItem) != -1) {
+                  int i = symbolItem->indexOfChild(currentItem);
+                  if (i > 0)
+                        treeWidget->setCurrentItem(symbolItem->child(i-1));
+                  else
+                      selectItem(symbolItem->text(0));
+                  }
+            else {
+                  selectItem(treeWidget->topLevelItem(treeWidget->topLevelItemCount() - 1)->text(0));
+                  }
+            }
+      else if (event->key() == Qt::Key_Right) {
+            QWidget* w = stack->currentWidget();
+            QWidget* ch = w->childAt(1, 1);
+            Palette* p = 0;
+            if (currentItem == keyItem) {
```

Comment:
Same comment about special casing here.  If it's the only way, so be it, but I don't understand why setting / checking attributes wouldn't work better.